scipy/__init__.py

```diff
@@ -145,11 +143,53 @@ def __dir__():
                       f" is required for this version of SciPy (detected "
                       f"version {__numpy_version__}",
                       UserWarning)
-
     del _pep440
 
     from scipy._lib._ccallback import LowLevelCallable
 
     from scipy._lib._testutils import PytestTester
     test = PytestTester(__name__)
     del PytestTester
+
+    submodules = [
+        'cluster',
+        'fft',
+        'fftpack',
+        'integrate',
+        'interpolate',
+        'io',
+        'linalg',
+        'misc',
+        'ndimage',
+        'odr',
+        'optimize',
+        'signal',
+        'sparse',
+        'spatial',
+        'special',
+        'stats'
+    ]
+
+    __all__ = submodules + [
+        'LowLevelCallable',
+        'test',
+        'show_config',
+        '__version__',
+        '__numpy_version__'
+    ]
+
+    def __dir__():
+        return __all__
+
+    import importlib as _importlib
+
+    def __getattr__(name):
+        if name in submodules:
+            return _importlib.import_module(f'scipy.{name}')
```

Comment:
Just to make sure I understand:

- the changes in the PR are in support of: https://scientific-python.org/specs/spec-0001/ ?
- this and the testing changes below suggest that we would now be able to do something like:

```python
import scipy
scipy.spatial.ConvexHull(points)
# no more: AttributeError: module 'scipy' has no attribute 'spatial'?
```

Or am I misunderstanding? Is this pretty much the full-blown implementation of the SPEC already, or just the tip of the iceberg?

Reply:
I wanted to make the minimally invasive changes that would allow what you show above with spatial.

So, this does not take lazy loading as far as skimage, where it is used for all subpackages and also the content of those subpackages. I was initially proposing that, but it sounded like a step too far for scipy from the feedback received.

Reply:
Sounds good and if that's help for implementing the full spec later I am 👍 IMO that should be the objective in the end.

But then we need to update the doc as well (how to use the API part) as we describe that a user should import submodules.

Reply:
I purposefully did not document this anywhere, because I think that is something that the SciPy developers do not have agreement on (i.e., how we should instruct new users to access the library). But happy to add docs once there is some consensus!

Reply:
Ok thanks for the confirmation.